test/parallel/test-worker-locks.js

```diff
@@ -0,0 +1,198 @@
+'use strict';
+const { mustCall, mustNotCall } = require('../common');
+const assert = require('assert');
+const { locks, Worker } = require('worker_threads');
+const { EventEmitter, once } = require('events');
+
+const tick = () => new Promise((resolve) => setImmediate(resolve));
+
+(async () => {
+  {
+    // Acquiring a lock and releasing it works.
+    assert.deepStrictEqual(await locks.query(), { held: [], pending: [] });
+    await locks.request('lock-name', mustCall(async (lock) => {
+      const snapshot = await locks.query();
+      assert.strictEqual(snapshot.pending.length, 0);
+      assert.strictEqual(snapshot.held.length, 1);
+      assert.strictEqual(snapshot.held[0].name, 'lock-name');
+      assert.strictEqual(snapshot.held[0].mode, 'exclusive');
+      assert.strictEqual(lock.name, 'lock-name');
+      assert.strictEqual(lock.mode, 'exclusive');
+    }));
+    assert.deepStrictEqual(await locks.query(), { held: [], pending: [] });
+  }
+
+  {
+    // Acquiring a lock will make other lock requests for the same name wait.
+    assert.deepStrictEqual(await locks.query(), { held: [], pending: [] });
+    let firstRequestDone = false;
+    let secondRequest;
+    let secondRequestDone = false;
+    await locks.request('lock-name', mustCall(async (lock) => {
+      secondRequest = locks.request('lock-name', mustCall(async (lock) => {
```

Comment:
At some point we might want to look into an option for making locks re-entrant, so that if a context attempts to acquire a lock within a context where it's already held, it still just works. Obviously that 's not an easy thing to get right but could be worth exploring later on.

Reply:
> At some point we might want to look into an option for making locks re-entrant, so that if a context attempts to acquire a lock within a context where it's already held, it still just works.

I’m not a fan of adding non-standard options at this point – I agree that this would make sense, but also in non-Node.js contexts, so maybe this could be brought up on the WICG level?

> Obviously that 's not an easy thing to get right but could be worth exploring later on.

I think all we’d need is a reference counter on the held lock.